framework/platform/unix/direct_window.cpp

```diff
@@ -244,76 +244,193 @@ VkSurfaceKHR DirectWindow::create_surface(Instance &instance)
+	// Find possible display config candidates
+	std::vector<Candidate> candidates = find_display_candidates(phys_dev, properties.mode);
+	if (candidates.empty())
+	{
+		LOGE("Direct-to-display: No compatible display candidates found");
+		return VK_NULL_HANDLE;
+	}
 
-	// Get the list of planes
-	uint32_t num_planes;
-	VK_CHECK(vkGetPhysicalDeviceDisplayPlanePropertiesKHR(phys_dev, &num_planes, nullptr));
+	// 'candidates' contains valid potential display configs. Find the best one.
+	// Look for the candidate with the closest resolution to our requested width & height.
+	uint32_t best_candidate = 0;
+	uint32_t wanted_area    = properties.extent.width * properties.extent.height;
+	uint32_t best_diff      = ~0;
 
-	if (num_planes == 0)
+	for (size_t c = 0; c < candidates.size(); c++)
 	{
-		LOGE("Direct-to-display: No display planes found");
-		return VK_NULL_HANDLE;
+		const Candidate &cand = candidates[c];
+
+		uint32_t area = cand.display_props.physicalResolution.width *
+		                cand.display_props.physicalResolution.height;
+
+		uint32_t diff = abs(static_cast<int32_t>(area) - static_cast<int32_t>(wanted_area));
+
+		if (diff < best_diff)
+		{
+			best_diff      = diff;
+			best_candidate = c;
+		}
 	}
 
-	std::vector<VkDisplayPlanePropertiesKHR> plane_properties(num_planes);
+	const Candidate &best = candidates[best_candidate];
 
-	VK_CHECK(vkGetPhysicalDeviceDisplayPlanePropertiesKHR(phys_dev, &num_planes, plane_properties.data()));
+	// Get the full display mode extent
+	full_extent.width  = best.mode.parameters.visibleRegion.width;
+	full_extent.height = best.mode.parameters.visibleRegion.height;
 
-	// Find a compatible plane index
-	uint32_t plane_index = find_compatible_plane(phys_dev, display, plane_properties);
-	if (plane_index == ~0U)
+	VkExtent2D image_extent;
+	if (properties.mode == Window::Mode::Fullscreen ||
+	    properties.mode == Window::Mode::FullscreenBorderless)
 	{
-		return VK_NULL_HANDLE;
+		// Fullscreen & Borderless options create a surface that matches the display size
+		image_extent.width  = full_extent.width;
+		image_extent.height = full_extent.height;
+	}
+	else
+	{
+		// Normal and stretch options create a surface at the requested width & height
+		// clamped to the plane capabilities in play
+		image_extent.width  = std::min(best.caps.maxSrcExtent.width,
+                                      std::max(properties.extent.width, best.caps.minSrcExtent.width));
+		image_extent.height = std::min(best.caps.maxSrcExtent.height,
+		                               std::max(properties.extent.height, best.caps.minSrcExtent.height));
 	}
 
-	VkExtent2D image_extent;
-	image_extent.width  = mode_props.parameters.visibleRegion.width;
-	image_extent.height = mode_props.parameters.visibleRegion.height;
+	// Calculate the display DPI
+	dpi = 25.4f * best.display_props.physicalResolution.width / best.display_props.physicalDimensions.width;
```

Comment:
Magic number `25.4f`? What is it?
Replace it with some define?

Reply:
Done